### ISSUE TITLE
fatal: your current branch 'master' does not have any commits yet

### ISSUE THREAD

**Issue by sfllaw** [sourcegraph/issues#78]
- Issue type: Bug report
- Sourcegraph version: 2.7.6 (15143_2018-05-09_c31b624) `sourcegraph/server:insiders`
- OS Version: amzn2-ami-hvm-2017.12.0.20180328.1-x86_64-gp2 (ami-07eb707f)
- Docker version:
```
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   3dfb8343b139d6342acfd9975d7f1068b5b1c3d3
 Built:        Tue Nov 14 22:03:51 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   402dd4a/17.06.2-ce
 Built:        Tue Nov 14 22:04:39 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

#### Steps to reproduce:
1. Create a completely empty repository called `github.example.com/user/empty-repo`
2. Enable that repository in Sourcegraph
3. Search for `type:diff repo:@*refs/heads/ \/\/\s*(TODO|BUG|FIXME)`

#### Expected response
No error messages and either search results, or a message saying that nothing was found. Empty repositories should not break search.

#### Actual result
An error message and no search results:
> 🛑 1 error occurred: \* diff search failed: failed to search commit diffs github.example.com/user/empty-repo@\*refs/heads/: exec git [git show --no-patch --no-merges -z --decorate=full --patch --format=format:%H%x00%D%x00%aN%x00%aE%x00%at%x00%cN%x00%cE%x00%ct%x00%B%x00%P%x00 --max-count=31 --unified=0 --no-prefix --extended-regexp --regexp-ignore-case --glob=refs/heads/ -G\\/\\/\\s\*(TODO|BUG|FIXME) --regexp-ignore-case --pickaxe-regex --extended-regexp --] command failed: exit status 128 (stderr: "fatal: your current branch 'master' does not have any commits yet"). Output was:

#### JSON response
```json
{
    "data": {
        "search": {
            "results": null
        }
    },
    "errors": [
        {
            "message": "1 error occurred:\n\n* diff search failed: failed to search commit diffs github.example.com/user/empty-repo@*refs/heads/: exec git [git show --no-patch --no-merges -z --decorate=full --patch --format=format:%H%x00%D%x00%aN%x00%aE%x00%at%x00%cN%x00%cE%x00%ct%x00%B%x00%P%x00 --max-count=31 --unified=0 --no-prefix --extended-regexp --regexp-ignore-case --glob=refs/heads/ -G\\/\\/\\s*(TODO|BUG|FIXME) --regexp-ignore-case --pickaxe-regex --extended-regexp --] command failed: exit status 128 (stderr: \"fatal: your current branch master does not have any commits yet\"). Output was:\n\n",
            "path": [
                "search",
                "results"
            ]
        }
    ]
}
```

#### Workaround
Disable empty repositories.

**Comment by nicksnyder** [comment]
Hmm, I can't seem to reproduce this.

I created a public empty repo: https://github.com/nicksnyder/empty and it is the only repo on my instance, and I see this:

![image](https://user-images.githubusercontent.com/754768/39940236-72c641d4-550d-11e8-8feb-6a2b25418b2c.png)

I also tried adding other repos and did not trigger the error you saw.

Some questions to help figure out what is different between our setups:
- Are you able to reproduce with my empty repo?
- What code host are you using and what steps did you take to create an empty repo on it?
- What is in your site config?

We have been fixing other bugs recently so another thing you could try is pulling the latest image again.

**Comment by sfllaw** [comment]
@nicksnyder I tested this on 15251_2018-05-11_14b0a8c and I can still reproduce it.

It looks like it's intermittent on our production system. I'll try to fire up our staging server and see if I can reproduce the error there.

We are using GitHub Enterprise, but it looks like the problem is that Sourcegraph is running `git show` on an empty repository:
```
/var/opt/sourcegraph/repos/github.example.com/user/empty-repo # git show
fatal: your current branch 'master' does not have any commits yet
```

**Comment by sfllaw** [comment]
Starting a staging server from scratch and cloning only empty repositories does not seem to trigger this bug. Although it is intermittent, so I might just be getting “lucky”.

I don't know how you’ve designed Sourcegraph, but does it keep track of empty repositories and skips running `git show` on them? If so, could it be that the `.git` directory is now out of date with Sourcegraph’s internal cache?

I can imagine that a repository might empty itself out if all its branches get deleted. Or if the repository itself gets deleted upstream.

**Comment by keegancsmith** [comment]
We don't just run `git show`, we pass extra args which changes its behaviour. IE even on an empty repository `git show` with a ref glob should work.

What happens when you run 

```
git show --no-patch --no-merges -z --decorate=full --patch --max-count=31 --unified=0 --no-prefix --extended-regexp --regexp-ignore-case --glob=refs/heads/ -G'\/\/\s*(TODO|BUG|FIXME)' --regexp-ignore-case --pickaxe-regex --extended-regexp --
```

on your empty repo?

**Comment by sfllaw** [comment]
```console
[ec2-user@ip-10-0-0-1 ~]$ docker exec -t -i sourcegraph /bin/sh
/ # cd /var/opt/sourcegraph/repos/github.example.com/user/repo/
/var/opt/sourcegraph/repos/github.example.com/user/repo # git show --no-patch --no-merges -z --decorate=full --patch --max-count=31 --unified=0 --no-pre
fix --extended-regexp --regexp-ignore-case --glob=refs/heads/ -G'\/\/\s*(TODO|BUG|FIXME)' --regexp-ignore-case --pickaxe-regex --extended-regexp --
fatal: your current branch 'master' does not have any commits yet
```

Are you sure it's a truly empty repository? I get the same message when creating a repo from scratch:
```console
/tmp # git init empty-repo
Initialized empty Git repository in /tmp/empty-repo/.git/
/tmp # cd empty-repo
/tmp/empty-repo # git show --no-patch --no-merges -z --decorate=full --patch --max-count=31 --unified=0 --no-prefix --extended-regexp --regexp-ignore-case --glob=refs/head
s/ -G'\/\/\s*(TODO|BUG|FIXME)' --regexp-ignore-case --pickaxe-regex --extended-regexp --
fatal: your current branch 'master' does not have any commits yet
```

**Comment by keegancsmith** [comment]
Bingo. Its a bug in git. I'm sorry, I must of not checked on a build of server which is silly of me. I checked on our dogfood instance and locally (I assumed dogfood had the same git version as server). But yeah this will be a source of issues @nicksnyder:

* dev `git version 2.17.0`
* server `git version 2.13.5`
* dogfood `git version 2.15.0`

We already have an existing issue about using an old version of alpine, but there are some complications to just bumping it. But this is more evidence its worth it. I suppose what we need to do now is see how easy it is to install a newer git and/or add a check for empty repos in diff search.

**Comment by sfllaw** [comment]
@keegancsmith Confirmed that git 2.15.0 from Alpine 3.7 fixes this bug.

**Comment by nicksnyder** [comment]
@sfllaw I updated the version of git we use in our latest build, which should fix the issue.

**Comment by sfllaw** [comment]
Confirmed fixed in 2.8.9.
